### PR TITLE
Nvcr 192 complete forgot password page

### PR DESCRIPTION
**Pull request by NadiiaNanivska**

### CODE REVIEW

**Review by Vitalik2705 — APPROVED** [review]

**Review by KateSydor — APPROVED** [review]

**Review by KateYatsiuk — APPROVED** [review]

**Review by OstapKH — APPROVED** [review]

**Review by dmitru4ok — APPROVED** [review]

File: invoicer_project/user/views.py [hunk 1]
```diff
@@ -65,6 +68,57 @@ def create(self, request):
...
+    @action(detail=True, methods=['post'])
+    def send_confirmation_code(self, request):
+        data = JSONParser().parse(request)
+        email = data.get('email')
+        try:
+            user = User.objects.get(email=email)
+        except User.DoesNotExist:
+            user = None
+        if user:
+            code = ''
```
Comment (maxchv) [inline invoicer_project/user/views.py:81]: `code` is always empty string! Is it correct?

File: invoicer_project/user/views.py [hunk 2]
```diff
@@ -65,6 +68,57 @@ def create(self, request):
...
+            while not (any(c.isalpha() for c in code) and any(c.isdigit() for c in code)):
```
Comment (maxchv) [inline invoicer_project/user/views.py:82]: this loop doesn't have any sense if `code` contains empty string

File: invoicer_project/user/views.py [hunk 3]
```diff
@@ -65,6 +68,57 @@ def create(self, request):
...
+    
+    @action(detail=True, methods=['post'])
+    def send_confirmation_code(self, request):
+        data = JSONParser().parse(request)
+        email = data.get('email')
+        try:
+            user = User.objects.get(email=email)
+        except User.DoesNotExist:
+            user = None
+        if user:
```
Comment (maxchv) [inline invoicer_project/user/views.py:80]:
```
if not user:
    return JsonResponse({'error': 'Email address not found'}, status=404)
code = ''
...
```
It is better to replace nested conditionals with flat code whenever possible.

File: invoicer_project/user/views.py [hunk 4]
```diff
@@ -65,6 +68,57 @@ def create(self, request):
...
+        
+    def partial_update(self, request):
+        data = JSONParser().parse(request)
+        try:
+            user = User.objects.get(email=data['email'])
+        except User.DoesNotExist:
+            user = None
+        data['id'] = user.id
+        serializer = UserSettingsSerializer(user, data=data, partial=True)
+        if serializer.is_valid():
```
Comment (maxchv) [inline invoicer_project/user/views.py:113]: It is also possible to invert the conditional here:
```
if not serializer.is_valid():
    return JsonResponse({'errors': serializer.errors}, status=400)
new_password = ...
```

**Review by maxchv — APPROVED** [review]

File: invoicer_project/user/views.py (hunk 1 above)
Comment (NadiiaNanivska) [reply invoicer_project/user/views.py:81]: Done

File: invoicer_project/user/views.py (hunk 3 above)
Comment (NadiiaNanivska) [reply invoicer_project/user/views.py:80]: Done

File: invoicer_project/user/views.py (hunk 4 above)
Comment (NadiiaNanivska) [reply invoicer_project/user/views.py:113]: Done

**Pull request merged**